gap/semigroups/semieunit.gi

```diff
@@ -0,0 +1,715 @@
+#############################################################################
+##
+#W  semieunit.gi
+#Y  Copyright (C) 2017                                    Christopher Russell
+##
+##  Licensing information can be found in the README file of this package.
+##
+#############################################################################
+#
+# This code is based on the description of McAlister triple semigroups found
+# in Fundamentals of Semigroup Theory by John Howie.
+#
+#############################################################################
+# Methods for creating McAlister triple semigroups
+#############################################################################
+InstallMethod(McAlisterTripleSemigroup,
+"for a perm group, action, digraph, and digraph",
+[IsGroup, IsFunction, IsDigraph, IsDigraph],
+function(G, act, X, Y)
+  local anti_act, hom, y_edges, out_nbrs, orbs, min, fam, filt, M, x, y;
+
+  if not IsFinite(G) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the first argument must be a finite group,");
+  fi;
+
+  anti_act := function(pt, g)
+    return act(pt, g ^ -1);
+  end;
+
+  hom := ActionHomomorphism(G, DigraphVertices(X), anti_act);
+
+  if not IsSubgroup(AutomorphismGroup(X), Image(hom)) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the first argument (a group) must act by order ",
+                  "automorphisms on the third argument (a partial order ",
+                  "digraph),");
+  elif not IsPartialOrderDigraph(X) then
+    ErrorNoReturn("Semigroups: McAlisterTripleSemigroup: usage,\n",
+                  "the third argument must be a partial order digraph,");
+  fi;
+
+  # To check if the edges of y are a subset of the edges of x we need to apply
+  # the labeling of y's vertices to its edges.
+  y_edges := List(DigraphEdges(Y), edge -> [DigraphVertexLabels(Y)[edge[1]],
```

Comment:
I agree this should be documented. I wrote it a long time ago and was told that the labelling system maybe changing in Digraphs so that I didn't have to hack edge labels like this.